### ISSUE TITLE
Poor ergonomics for shortcuts

### ISSUE THREAD

**Issue by slikts** [ajtoo/vscode-org-mode#118]
I'm used to being able to create outlines just by using Enter at the end of line to insert a sibling item (or remove the item if it's empty) and Tab and Shift+Tab at the beginning of line to promote and demote items; C-M-o p/d/s for the same has the advantage that it works with the cursor anywhere in the line, but still is cumbersome for common repeated actions.

**Comment by sipi** [comment]
Concretely, what would you like to see implemented ?
Because you can already change keybinding for "do promote/demote"… Ok, it will be difficult to use "Tab" without blocking its behavior elsewhere in the file but I think we can't have a different behavior for a key based on the cursor context (but I can be wrong)…

We can add a command to insert a sibling outline "insert heading"… You will be able to define "CTRL+ENTER" to create new sibling outline like in default Emacs behavior…

**Comment by slikts** [comment]
I'm not immediately sure what's the best for this specific case, but the scheme I described is reasonably comfortable. What's clear is that <kbd>C-M-o p/d/s</kbd> is not comfortable for operations that are used *all* the time.

**Comment by sipi** [comment]
@KyrychukD yes, you can write a command that inspects the current cursor position and do different things depending on the context but you overwrites the default behavior. Especially for `enter`, it does not create new line anymore except if you recode this default behavior which is not really a very nice way to do. The shortcut handler should implements a "chain of responsability pattern" to do that. Which to my knowledge is not the case.

**Comment by slikts** [comment]
It shouldn't be the main concern for shortcuts to just not conflict; there should be a default scheme that has reasonable ergonomics and the option for users to customize it. 

As an example, there's a markdown extension that has the list editing features I described: https://github.com/neilsustc/vscode-markdown

**Comment by KyrychukD** [comment]
@sipi I see your point. I do not insist on changing the default behavior. However, I do not think it's a bad idea either. It makes sense that certain actions lead to slightly different results depending on context. I can come up with three more examples where `enter` does not only insert a newline character besides the extension that @slikts mentioned:

1. In vscode when you hit `enter` while the cursor is on an indented line, new line will be indented as well.
2. In vscode when you hit `enter` while the cursor is within empty brackets or parentheses, in addition to the new line with the cursor, the closing bracket or parenthesis to another new line.
3. When you're making a markdown list in Slack and you hit `enter` at the end of the line, a new item will be created (`- ` or `{number}. ` inserted depending the kind of the list) on the next line.

In all these cases the `enter` behavior is not the *default* behavior that you'd expect from `notepad.exe`. But I think that this context-dependent non-default behavior is convenient.

BTW, this hypothetical `magicHeadings` command does not have to be included in this extension. It is totally possible to put it in a separate extension as long as `org.insertSubheading` and `org.insertHeadingRespectContent` commands are exposed.

**Comment by sipi** [comment]
First of all, from my point of view, I agree with a context dependent behavior of some key like `enter`.  It's just a technical issue about, can we do it ? how to do it right ?

@KyrychukD for your first two points, I think they are hardcoded in vscode. The both works even in a .txt file.
But, ok I tested the markdown extension. And, I agree that it manages well `enter` key for new item list… However, in fact they recode the default `enter` behavior. Below, the last `else` of the command (this command say: I typed a '\n' character from my keyboard):
```
} else {
                return commands.executeCommand('type', { source: 'keyboard', text: '\n' });
}
```
I'm not really happy with that because if now I rebind this command to `TAB` for example, **`TAB` will insert new line** outside a list context :/ !!!

The right way to do this would be to be able to define the context in which the command can act, and if it can't, pass the baton. Which is the "chain of responsability pattern". So, if the "insert new item list command" detect, it is not in a list context then it return false (for exemple) and default behavior of the entered key occurs. In this way, `TAB` will not insert new line which is an undesirable side effect, and will insert indentation as expected.

**Comment by KyrychukD** [comment]
I am glad that we reached an agreement. Let's continue discussing details.

> I'm not really happy with that because if now I rebind this command to TAB for example, TAB will insert new line outside a list context :/ !!!

Correct. But I would simply call this a misuse of a feature that was designed specifically to enhance the behavior of the `enter` key. Nobody would map this command to TAB if the command documentation would say "Insert sibling heading below if cursor is on the end of a heading line, insert a new line otherwise". If they do, that's their choice. But why would they do it if they can *already* bind the command that creates a sibling heading to any key with the added benefit that the existing `org.insertHeadingRespectContent` command works regardless of the cursor position within a line.

Also, today I learned that you can reference settings in `when` part of the keybindings, which makes something like this possible:

```json
    {
        "key": "enter",
        "command": "org.magicHeadings",
        "when": "editorTextFocus && editorLangId == 'org' && org.magicHeadings.enabled"
    }
```
So that users could disable the keybinding completely by adding `org.magicHeadings.enabled: false` to their `settings.json`.

> The right way to do this would be to be able to define the context in which the command can act, and if it can't, pass the baton. Which is the "chain of responsability pattern". So, if the "insert new item list command" detect, it is not in a list context then it return false (for exemple) and default behavior of the entered key occurs.

I am not aware of a way to implement this as you describe in vscode. Once vscode matches a keybinding to a command, it stops looking, there is no way to affect this from within the command. I think the best we can do is use settings in `when` condition to restrict the context in which the command can be matched to a minimum as shown above.
